sw/device/tests/sim_dv/pwrmgr_deep_sleep_all_wake_ups.c

```diff
@@ -0,0 +1,294 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_adc_ctrl.h"
+#include "sw/device/lib/dif/dif_pinmux.h"
+#include "sw/device/lib/dif/dif_pwrmgr.h"
+#include "sw/device/lib/dif/dif_rv_plic.h"
+#include "sw/device/lib/dif/dif_sysrst_ctrl.h"
+#include "sw/device/lib/irq.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/pwrmgr_testutils.h"
+#include "sw/device/lib/testing/rv_plic_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "sw/device/lib/testing/autogen/isr_testutils.h"
+/*
+  PWRMGR DEEP SLEEP ALL WAKE UPS test
+
+  This test runs power manager wake up from deep leep mode by
+  wake up inputs.
+
+  There are 6 wake up inputs.
+  0: sysrst_ctrl
+  1: adc_ctrl
+  2: pinmux
+  3: usb
+  4: aon_timer
+  5: sensor_ctrl
+
+  #3 and 4 are covered by separate test (see
+  https://github.com/lowRISC/opentitan/blob/master/
+  hw/top_earlgrey/data/chip_testplan.hjson)
+
+  #5 is excluded because sensor_ctrl is not aon domain
+
+ */
+#define kTestNumWkup 3
+#define kNoWkupDev 0
+#define kPinmuxWkupDetector5 5
+
+#define MAKE_INIT_FUNC(dif_)                                             \
+  void dif_##_init(void *dif, uintptr_t base) {                          \
+    CHECK_DIF_OK(dif##_##dif_##_init(mmio_region_from_addr(base), dif)); \
+  }
+
+MAKE_INIT_FUNC(pwrmgr);
+MAKE_INIT_FUNC(rv_plic);
+MAKE_INIT_FUNC(sysrst_ctrl);
+MAKE_INIT_FUNC(adc_ctrl);
+MAKE_INIT_FUNC(pinmux);
+
+static dif_pwrmgr_t pwrmgr;
+static dif_rv_plic_t rv_plic;
+static dif_sensor_ctrl_t sysrst_ctrl;
+static dif_adc_ctrl_t adc_ctrl;
+static dif_pinmux_t pinmux;
+
+typedef struct test_dev {
+  /**
+   * Name of the device.
+   */
+  const char *name;
+  /**
+   * Base address for the device.
+   */
+  uintptr_t base;
+  /**
+   * Handle to the DIF object for this device.
+   */
+  void *dif;
+  /**
+   * Initialization for the DIF based on an address location.
+   *
+   * May be `NULL`.
+   */
+  void (*init)(void *dif, uintptr_t base);
+  /**
+   * Wakeup Sources
+   */
+  dif_pwrmgr_request_sources_t wakeup_src;
+  /**
+   * Configuration and initialization actions for the device.
+   * This will be passed the value of `dif` above.
+   */
+  void (*config)(void *dif);
+} test_dev_t;
+
+enum {
+  // test wake up input 0 : sysrst_ctrl
+  kRound1 = 0,
+  // test wake up input 1 : adc_ctrl
+  kRound2 = 1,
+  // test wake up input 2 : pinmux wakeup detector
+  kRound3 = 2,
+};
+
+/**
+ * sysrst_ctrl config for test #1
+ * . set sysrst_ctrl.KEY_INTR_CTL.pwrb_in_H2L to 1
+ * . use IOR13 as pwrb_in
+ */
+static void prgm_sysrst_ctrl_wakeup(void *dif) {
```

Comment:
Why does this need to be void*? Moreover, why do you need to pass it in? its declared globally visible within this file, no? (applies throughout to similar helper functions)

Reply:
same as above comments, all these wrapper functions are to apply in the loop with a function pointer.
Since participants of this test are relatively large, put all these are in one place, kTestWakeupSources[]
then call with a same routine.

Reply:
Ah I see. thanks for clarifying.